src/System.Private.CoreLib/shared/System/Globalization/ISOWeek.cs

```diff
@@ -0,0 +1,135 @@
+        private const int MaxWeek = WeeksInLongYear;
+
+        public static int GetWeekOfYear(DateTime date)
+        {
+            int week = GetWeekNumber(date);
+
+            if (week < MinWeek)
+            {
+                return GetWeeksInYear(date.Year - 1);
+            }
+
+            if (week > GetWeeksInYear(date.Year))
+            {
+                return 1;
+            }
+
+            return week;
+        }
+
+        public static int GetYear(DateTime date)
+        {
+            int week = GetWeekNumber(date);
+
+            if (week < MinWeek)
+            {
+                return date.Year - 1;
+            }
+
+            if (week > GetWeeksInYear(date.Year))
+            {
+                return date.Year + 1;
+            }
+
+            return date.Year;
+        }
+
+        public static DateTime GetYearStart(int year)
+        {
+            return ToDateTime(year, MinWeek, DayOfWeek.Monday);
+        }
+
+        public static DateTime GetYearEnd(int year)
+        {
+            return ToDateTime(year, GetWeeksInYear(year), DayOfWeek.Sunday);
+        }
+
+        // From https://en.wikipedia.org/wiki/ISO_week_date#Weeks_per_year.
+        public static int GetWeeksInYear(int year)
+        {
+            if (year < MinYear || year > MaxYear)
+            {
+                throw new ArgumentOutOfRangeException(nameof(year), SR.ArgumentOutOfRange_Year);
+            }
+
+            int P(int y) => (y + (y / 4) - (y / 100) + (y / 400)) % 7;
+
+            if (P(year) == 4 || P(year - 1) == 3)
+            {
+                return WeeksInLongYear;
+            }
+
+            return WeeksInShortYear;
+        }
+
+        // From https://en.wikipedia.org/wiki/ISO_week_date#Calculating_a_date_given_the_year,_week_number_and_weekday:
+        //
+        // This method requires that one know the weekday of 4 January of the year in question.
+        // Add 3 to the number of this weekday, giving a correction to be used for dates within this year.
+        //
+        // Multiply the week number by 7, then add the weekday. From this sum subtract the correction for the year.
+        // The result is the ordinal date, which can be converted into a calendar date.
+        //
+        // If the ordinal date thus obtained is zero or negative, the date belongs to the previous calendar year.
+        // If greater than the number of days in the year, to the following year.
+        public static DateTime ToDateTime(int year, int week, DayOfWeek dayOfWeek)
+        {
+            if (year < MinYear || year > MaxYear)
+            {
+                throw new ArgumentOutOfRangeException(nameof(year), SR.ArgumentOutOfRange_Year);
+            }
+
+            if (week < MinWeek || week > MaxWeek)
+            {
+                throw new ArgumentOutOfRangeException(nameof(week), SR.ArgumentOutOfRange_Week_ISO);
+            }
+
+            if ((int)dayOfWeek < 0 || (int)dayOfWeek > 6)
```

Comment:
> if ((int)dayOfWeek < 0 || (int)dayOfWeek > 6) [](start = 11, length = 46)

should we support the value 7 for Sunday too? I mean 0 and 7 is Sunday. 

I can see the scenario of someone getting the data from some source and just case the value to DayOfWeek.

Reply:
I'm not sure. We won't be able to reuse the existing exception message (`ArgumentOutOfRange_DayOfWeek`):

> The DayOfWeek enumeration must be in the range 0 through 6.

Reply:
> I'm not sure. We won't be able to reuse the existing exception message (ArgumentOutOfRange_DayOfWeek):
Why? what I am trying to Say, if someone passed dayOfWeek = 7 to us, we'll treat it as Sunday. so we'll not throw. I don't think this affects the exception thrown for other cases. right?

Reply:
And we can document the behavior. also, we can have a new exception message as needed too.

Reply:
Alright. Let's also accept 7 😄